### PR TITLE
RJR - support paragraphs in banner body copy

### PR DESCRIPTION
**Pull request by KaliedaRik**
## What does this change?
All Banners currently use a `messageText` String field to hold the copy to be used in the Banner's body. Different Banners also use different approaches to adding a sentence of highlighted text to the body copy. Compare this to the situation in Epics, where body copy is supplied in an array of Strings (in a `paragraphs` attribute), which allows Epics to break their body copy into paragraphs.

This PR aims to bring the handling of body copy into alignment between Banners and Epics, with copy supplied in a new `paragraphs` array. It also introduces a shared function to handle the addition of highlighted text to the last paragraph of the body copy.

### Risks
Body copy is generated by Marketing colleagues using the RRCP. This PR is supported by [an equivalent PR in the support.dotcom.components repo](https://github.com/guardian/support-dotcom-components/pull/632) to update the Banner variant forms so that they save their body content as an array of Strings in the new `paragraphs` attribute. (link to SAC PR required)

A major risk for this PR is handling the transition of existing Banner tests from the old way of doing things (using `messageText`) to the new way (using `paragraphs`). We mitigate that risk in this PR by including temporary functionality so that Banner tests that do not include the `paragraphs` copy automatically fall back to using the `messageText` copy.

Once we are satisfied that all Banner tests on the LIVE site have been updated to include their copy in the `paragraphs` attribute we will be able to remove the temporary functionality and simplify the code base.

### Testing
This PR can be tested using the RRCP in DEV or CODE.

### Images
(Add some images to compare existing vs new paragraph functionality)

### CODE REVIEW

File: public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx [hunk 1]
```diff
@@ -23,7 +23,10 @@ import TestVariantsSplitEditor from '../testVariantsSplitEditor';
 import { useStyles } from '../helpers/testEditorStyles';
 
 const copyHasTemplate = (content: BannerContent, template: string): boolean =>
-  (content.heading && content.heading.includes(template)) || content.messageText.includes(template);
+  (content.heading && content.heading.includes(template)) ||
+  (content.paragraphs != null && content.paragraphs[0].includes(template)) ||
```
Comment (KaliedaRik) [inline public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx:27]: Not happy with this line - I think that we need to be testing for the template variable against all paragraphs joined into a string?

File: public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx [hunk 2]
```diff
@@ -184,10 +199,14 @@ const BannerTestVariantContentEditor: React.FC<BannerTestVariantContentEditorPro
               required: EMPTY_ERROR_HELPER_TEXT,
               validate: templateValidator,
             })}
-            error={errors.messageText !== undefined}
-            helperText={errors.messageText ? errors.messageText.message : BODY_DEFAULT_HELPER_TEXT}
+            error={errors.paragraphs !== undefined}
+            helperText={
+              errors.paragraphs && errors.paragraphs[0]
```
Comment (KaliedaRik) [inline public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx:204]: Another 'not happy with' comment - this time I'm not sure about how the error checker records error messages for an array of copy. Need to check with how it works with the Epic paragraphs array code in main branch.

File: public/src/components/channelManagement/bannerTests/utils/defaults.ts [hunk 3]
```diff
@@ -30,6 +33,7 @@ const PROD_DEFAULT_VARIANT: BannerVariant = {
   template: BannerTemplate.ContributionsBanner,
   bannerContent: {
     messageText: '',
+    paragraphs: [''],
```
Comment (KaliedaRik) [inline public/src/components/channelManagement/bannerTests/utils/defaults.ts:36]: Need to check to see if setting paragraphs to an empty array would be the better thing to do here.

File: public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx (hunk 1 above)
Comment (tomrf1) [reply public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx:27]: Just check all paras in the array -
`(content.paragraphs && content.paragraphs.some(para => para.includes(template))`

File: public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx (hunk 2 above)
Comment (tomrf1) [reply public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx:204]: For the epic it's actually maintained as a string up until `onSubmit` - see [here](https://github.com/guardian/support-admin-console/blob/main/public/src/components/channelManagement/epicTests/epicTestVariantEditor.tsx#L119).

We'll change that when the new RichTextEditor component is added, but for now I suggest following what the epic does. This unfortunately involves having a [separate type](https://github.com/guardian/support-admin-console/blob/main/public/src/components/channelManagement/epicTests/epicTestVariantEditor.tsx#L58) for the `useForm`.

File: public/src/components/channelManagement/bannerTests/utils/defaults.ts (hunk 3 above)
Comment (tomrf1) [reply public/src/components/channelManagement/bannerTests/utils/defaults.ts:36]: I would go with empty array, and remove `messageText` entirely

Comment (KaliedaRik) [reply public/src/components/channelManagement/bannerTests/utils/defaults.ts]: Done

File: public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx (hunk 2 above)
Comment (KaliedaRik) [reply public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx]: Banner body copy now follows Epic functionality

File: public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx (hunk 1 above)
Comment (KaliedaRik) [reply public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx]: Done

File: public/src/components/channelManagement/bannerTests/utils/defaults.ts [hunk 4]
```diff
@@ -30,6 +33,7 @@ const PROD_DEFAULT_VARIANT: BannerVariant = {
   template: BannerTemplate.ContributionsBanner,
   bannerContent: {
     messageText: '',
```
Comment (tomrf1) [inline public/src/components/channelManagement/bannerTests/utils/defaults.ts:35]: `messageText` needs to be removed from the defaults now

Comment (KaliedaRik) [reply public/src/components/channelManagement/bannerTests/utils/defaults.ts:35]: Done!

File: public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx [hunk 5]
```diff
@@ -127,10 +146,10 @@ const BannerTestVariantContentEditor: React.FC<BannerTestVariantContentEditorPro
   useEffect(() => {
     const isValid = Object.keys(errors).length === 0;
     onValidationChange(isValid);
-  }, [errors.messageText, errors.heading, errors.highlightedText]);
+  }, [errors.body, errors.heading, errors.highlightedText]);
 
-  const onSubmit = ({ heading, messageText, highlightedText }: BannerContent): void => {
-    onChange({ ...content, heading, messageText, highlightedText });
+  const onSubmit = ({ heading, body, highlightedText }: FormData): void => {
+    onChange({ ...content, heading, paragraphs: body.split('\n'), highlightedText });
```
[the remark this answers is not in the document]
Comment (KaliedaRik) [reply public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx:152]: Would this change lead to the removal of cta and secondaryCta data each time someone updates the copy fields? An alternative approach to getting rid of `messageText` may be to explicitly (and temporarily) set it to null?

[the remark this answers is not in the document]
Comment (tomrf1) [reply public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx:152]: No - this component updates the content only, see the type of `onChange` -
`onChange: (updatedContent: BannerContent) => void;`

Also, TS would tell us if we were removing any required fields.

[the remark this answers is not in the document]
Comment (tomrf1) [reply public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx:152]: oh sorry, just realised ctas are on the content model!

[the remark this answers is not in the document]
Comment (tomrf1) [reply public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx:152]: ok in that case I agree with explicitly removing messageText (for now)

[the remark this answers is not in the document]
Comment (tomrf1) [reply public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx:152]: side note - it is odd that `onSubmit` doesn't handle updates to the ctas, but let's not worry about that now!

[the remark this answers is not in the document]
Comment (KaliedaRik) [reply public/src/components/channelManagement/bannerTests/bannerTestVariantEditor.tsx]: Done

**Review by tomrf1 — APPROVED** [review]
great work!

**Pull request merged**